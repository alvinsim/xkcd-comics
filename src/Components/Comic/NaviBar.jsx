import { Button, Icon } from 'antd';
import PropTypes from 'prop-types';
import React from 'react';
import { withRouter } from 'react-router';

const NaviBar = (props) => {
  const isDisableFirstButton = () => {
    const { first, current } = props;

    return first >= current;
  };

  const isDisablePreviousButton = () => {
    return isDisableFirstButton();
  }

  const isDisableNextButton = () => {
    const { current, latest } = props;

    return current === latest;
  }

  const isDisableLatestButton = () => {
    return isDisableNextButton();
  }

  const onClickFavourites = () => {
    const { history } = props;
    history.push('/favourites');
  }

  const {
    onClickFirst,
    onClickLatest,
    onClickNext,
    onClickPrevious,
    onClickRandom
  } = props;

  return (
    <div className="navi-bar">
      <Button.Group>
        <Button
          disabled={isDisableFirstButton()}
          onClick={onClickFirst}
          title="first"
        >
          <Icon type="double-left" />
        </Button>
        <Button
          disabled={isDisablePreviousButton()}
          onClick={onClickPrevious}
          title="previous"
        >
          <Icon type="left" />
        </Button>
        <Button title="random" onClick={onClickRandom}>
          <Icon type="question" />
        </Button>
        <Button onClick={onClickFavourites} title="favourites">
          <Icon theme="twoTone" type="heart" />
        </Button>
        <Button
          disabled={isDisableNextButton()}
          onClick={onClickNext}
          title="next"
        >
          <Icon type="right" />
        </Button>
        <Button
          disabled={isDisableLatestButton()}
          onClick={onClickLatest}
          title="latest"
        >
          <Icon type="double-right" />
        </Button>
      </Button.Group>
    </div>
  );
};

NaviBar.propTypes = {
  current: PropTypes.number,
  first: PropTypes.number,
  history: PropTypes.instanceOf(Object),
  latest: PropTypes.number,
  onClickFirst: PropTypes.func.isRequired,
  onClickPrevious: PropTypes.func.isRequired,
  onClickNext: PropTypes.func.isRequired,
  onClickLatest: PropTypes.func.isRequired,
  onClickRandom: PropTypes.func.isRequired,
};

NaviBar.defaultProps = {
  current: 0,
  first: 1,
  history: {},
  latest: 0,
};

export default withRouter(NaviBar);
