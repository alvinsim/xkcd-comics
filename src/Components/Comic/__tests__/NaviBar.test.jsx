import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import NaviBar from '../NaviBar';

test('Next and Latest buttons are disabled when current comic is the latest', () => {
  const component = renderer.create(
    <BrowserRouter>
      <NaviBar
        current={5}
        latest={5}
        onClickFirst={jest.fn()}
        onClickPrevious={jest.fn()}
        onClickNext={jest.fn()}
        onClickLatest={jest.fn()}
        onClickRandom={jest.fn()}
      />
    </BrowserRouter>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Last and Previous buttons are disabled when current is the first comic', () => {
  const component = renderer.create(
    <BrowserRouter>
      <NaviBar
        current={1}
        latest={5}
        onClickFirst={jest.fn()}
        onClickPrevious={jest.fn()}
        onClickNext={jest.fn()}
        onClickLatest={jest.fn()}
        onClickRandom={jest.fn()}
      />
    </BrowserRouter>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})
