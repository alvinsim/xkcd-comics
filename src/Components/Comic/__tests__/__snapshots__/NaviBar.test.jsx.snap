// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Last and Previous buttons are disabled when current is the first comic 1`] = `
<div
  className="navi-bar"
>
  <div
    className="ant-btn-group"
  >
    <button
      className="ant-btn"
      disabled={true}
      onClick={[Function]}
      title="first"
      type="button"
    >
      <i
        aria-label="icon: double-left"
        className="anticon anticon-double-left"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="double-left"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M272.9 512l265.4-339.1c4.1-5.2.4-12.9-6.3-12.9h-77.3c-4.9 0-9.6 2.3-12.6 6.1L186.8 492.3a31.99 31.99 0 0 0 0 39.5l255.3 326.1c3 3.9 7.7 6.1 12.6 6.1H532c6.7 0 10.4-7.7 6.3-12.9L272.9 512zm304 0l265.4-339.1c4.1-5.2.4-12.9-6.3-12.9h-77.3c-4.9 0-9.6 2.3-12.6 6.1L490.8 492.3a31.99 31.99 0 0 0 0 39.5l255.3 326.1c3 3.9 7.7 6.1 12.6 6.1H836c6.7 0 10.4-7.7 6.3-12.9L576.9 512z"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      disabled={true}
      onClick={[Function]}
      title="previous"
      type="button"
    >
      <i
        aria-label="icon: left"
        className="anticon anticon-left"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="left"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 0 0 0 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      onClick={[Function]}
      title="random"
      type="button"
    >
      <i
        aria-label="icon: question"
        className="anticon anticon-question"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="question"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M764 280.9c-14-30.6-33.9-58.1-59.3-81.6C653.1 151.4 584.6 125 512 125s-141.1 26.4-192.7 74.2c-25.4 23.6-45.3 51-59.3 81.7-14.6 32-22 65.9-22 100.9v27c0 6.2 5 11.2 11.2 11.2h54c6.2 0 11.2-5 11.2-11.2v-27c0-99.5 88.6-180.4 197.6-180.4s197.6 80.9 197.6 180.4c0 40.8-14.5 79.2-42 111.2-27.2 31.7-65.6 54.4-108.1 64-24.3 5.5-46.2 19.2-61.7 38.8a110.85 110.85 0 0 0-23.9 68.6v31.4c0 6.2 5 11.2 11.2 11.2h54c6.2 0 11.2-5 11.2-11.2v-31.4c0-15.7 10.9-29.5 26-32.9 58.4-13.2 111.4-44.7 149.3-88.7 19.1-22.3 34-47.1 44.3-74 10.7-27.9 16.1-57.2 16.1-87 0-35-7.4-69-22-100.9zM512 787c-30.9 0-56 25.1-56 56s25.1 56 56 56 56-25.1 56-56-25.1-56-56-56z"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      onClick={[Function]}
      title="favourites"
      type="button"
    >
      <i
        aria-label="icon: heart"
        className="anticon anticon-heart"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="heart"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9zM512 814.8S156 586.7 156 385.5C156 283.6 240.3 201 344.3 201c73.1 0 136.5 40.8 167.7 100.4C543.2 241.8 606.6 201 679.7 201c104 0 188.3 82.6 188.3 184.5 0 201.2-356 429.3-356 429.3z"
            fill="#1890ff"
          />
          <path
            d="M679.7 201c-73.1 0-136.5 40.8-167.7 100.4C480.8 241.8 417.4 201 344.3 201c-104 0-188.3 82.6-188.3 184.5 0 201.2 356 429.3 356 429.3s356-228.1 356-429.3C868 283.6 783.7 201 679.7 201z"
            fill="#e6f7ff"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      disabled={false}
      onClick={[Function]}
      title="next"
      type="button"
    >
      <i
        aria-label="icon: right"
        className="anticon anticon-right"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      disabled={false}
      onClick={[Function]}
      title="latest"
      type="button"
    >
      <i
        aria-label="icon: double-right"
        className="anticon anticon-double-right"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="double-right"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M533.2 492.3L277.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H188c-6.7 0-10.4 7.7-6.3 12.9L447.1 512 181.7 851.1A7.98 7.98 0 0 0 188 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5zm304 0L581.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H492c-6.7 0-10.4 7.7-6.3 12.9L751.1 512 485.7 851.1A7.98 7.98 0 0 0 492 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5z"
          />
        </svg>
      </i>
    </button>
  </div>
</div>
`;

exports[`Next and Latest buttons are disabled when current comic is the latest 1`] = `
<div
  className="navi-bar"
>
  <div
    className="ant-btn-group"
  >
    <button
      className="ant-btn"
      disabled={false}
      onClick={[Function]}
      title="first"
      type="button"
    >
      <i
        aria-label="icon: double-left"
        className="anticon anticon-double-left"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="double-left"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M272.9 512l265.4-339.1c4.1-5.2.4-12.9-6.3-12.9h-77.3c-4.9 0-9.6 2.3-12.6 6.1L186.8 492.3a31.99 31.99 0 0 0 0 39.5l255.3 326.1c3 3.9 7.7 6.1 12.6 6.1H532c6.7 0 10.4-7.7 6.3-12.9L272.9 512zm304 0l265.4-339.1c4.1-5.2.4-12.9-6.3-12.9h-77.3c-4.9 0-9.6 2.3-12.6 6.1L490.8 492.3a31.99 31.99 0 0 0 0 39.5l255.3 326.1c3 3.9 7.7 6.1 12.6 6.1H836c6.7 0 10.4-7.7 6.3-12.9L576.9 512z"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      disabled={false}
      onClick={[Function]}
      title="previous"
      type="button"
    >
      <i
        aria-label="icon: left"
        className="anticon anticon-left"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="left"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 0 0 0 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      onClick={[Function]}
      title="random"
      type="button"
    >
      <i
        aria-label="icon: question"
        className="anticon anticon-question"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="question"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M764 280.9c-14-30.6-33.9-58.1-59.3-81.6C653.1 151.4 584.6 125 512 125s-141.1 26.4-192.7 74.2c-25.4 23.6-45.3 51-59.3 81.7-14.6 32-22 65.9-22 100.9v27c0 6.2 5 11.2 11.2 11.2h54c6.2 0 11.2-5 11.2-11.2v-27c0-99.5 88.6-180.4 197.6-180.4s197.6 80.9 197.6 180.4c0 40.8-14.5 79.2-42 111.2-27.2 31.7-65.6 54.4-108.1 64-24.3 5.5-46.2 19.2-61.7 38.8a110.85 110.85 0 0 0-23.9 68.6v31.4c0 6.2 5 11.2 11.2 11.2h54c6.2 0 11.2-5 11.2-11.2v-31.4c0-15.7 10.9-29.5 26-32.9 58.4-13.2 111.4-44.7 149.3-88.7 19.1-22.3 34-47.1 44.3-74 10.7-27.9 16.1-57.2 16.1-87 0-35-7.4-69-22-100.9zM512 787c-30.9 0-56 25.1-56 56s25.1 56 56 56 56-25.1 56-56-25.1-56-56-56z"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      onClick={[Function]}
      title="favourites"
      type="button"
    >
      <i
        aria-label="icon: heart"
        className="anticon anticon-heart"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="heart"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9zM512 814.8S156 586.7 156 385.5C156 283.6 240.3 201 344.3 201c73.1 0 136.5 40.8 167.7 100.4C543.2 241.8 606.6 201 679.7 201c104 0 188.3 82.6 188.3 184.5 0 201.2-356 429.3-356 429.3z"
            fill="#1890ff"
          />
          <path
            d="M679.7 201c-73.1 0-136.5 40.8-167.7 100.4C480.8 241.8 417.4 201 344.3 201c-104 0-188.3 82.6-188.3 184.5 0 201.2 356 429.3 356 429.3s356-228.1 356-429.3C868 283.6 783.7 201 679.7 201z"
            fill="#e6f7ff"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      disabled={true}
      onClick={[Function]}
      title="next"
      type="button"
    >
      <i
        aria-label="icon: right"
        className="anticon anticon-right"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn"
      disabled={true}
      onClick={[Function]}
      title="latest"
      type="button"
    >
      <i
        aria-label="icon: double-right"
        className="anticon anticon-double-right"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="double-right"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M533.2 492.3L277.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H188c-6.7 0-10.4 7.7-6.3 12.9L447.1 512 181.7 851.1A7.98 7.98 0 0 0 188 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5zm304 0L581.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H492c-6.7 0-10.4 7.7-6.3 12.9L751.1 512 485.7 851.1A7.98 7.98 0 0 0 492 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5z"
          />
        </svg>
      </i>
    </button>
  </div>
</div>
`;
